src/python/pants/core/goals/update_build_files.py

```diff
@@ -192,7 +211,7 @@ async def update_build_files(
 
     rewrite_request_classes = []
     for request in union_membership[RewrittenBuildFileRequest]:
-        if issubclass(request, FormatWithBlackRequest):
+        if issubclass(request, (FormatWithBlackRequest, FormatWithYapfRequest)):
             if update_build_files_subsystem.fmt:
                 rewrite_request_classes.append(request)
```

Comment:
This will need to consume the `.formatter` option. If formatter is set to Black, skip Yapf. Vice versa.

Reply:
Thanks for the hint, Eric. I am sorry, I don't follow. 

```
            if update_build_files_subsystem.fmt:
                if update_build_files_subsystem.formatter == Formatter.BLACK:
```

Why would we need to check the formatter? I was under impression that `update_build_files` would be invoked with a single `UpdateBuildFilesSubsystem`? 

And since I do need to consume the `.formatter` option, could you please give a hint at what should be modified? :/

Reply:
Sure! You are registering via `UnionRule` both the Black format request and the Yapf format request. So naively, Pants will try running both. It will pipe the first result to the next one.

I think you can rewrite the whole for loop like this:

```python
    for request in union_membership[RewrittenBuildFileRequest]:
        if issubclass(request, (FormatWithBlackRequest, FormatWithYapfRequest)):
            is_chosen_formatter = (
                issubclass(request, FormatWithBlackRequest) ^ update_build_files_subsytem.formatter == Formatter.YAPF
            )
            if update_build_files_subsystem.fmt and is_chosen_formatter:
                rewrite_request_classes.append(request)
            else:
                continue
        if update_build_files_subsystem.fix_safe_deprecations or not issubclass(
            request, DeprecationFixerRequest
        ):
            rewrite_request_classes.append(request)
```

That `^` is an "exclusive or". You're checking first that the formatter is Black; if it's not, we can assume it is Yapf because that is the only other option. Then, you're also checking whether the user told you to use Yapf. This check makes sure that both the class & the option correspond to each other.

Does that make sense? The `^` exclusive or is subtle, might be worth making the logic less "clever". What we care about is making sure that the `request` class corresponds to the `--formatter` option.

Reply:
Ah got it. I think XOR would be a lot less verbose than:

```
        if issubclass(request, (FormatWithBlackRequest, FormatWithYapfRequest)):
            if issubclass(request, FormatWithBlackRequest) and update_build_files_subsystem.formatter != Formatter.YAPF:
                is_chosen_formatter = True
            elif not issubclass(request, FormatWithBlackRequest) and update_build_files_subsystem.formatter == Formatter.YAPF:
                is_chosen_formatter = True
            else:
                is_chosen_formatter = False
```